### DISCUSSION TITLE
How to lazy load Policy method results. What is best practise?

### DISCUSSION THREAD

**Discussion by bojmaliev** [inertiajs/inertia#1085] [Help]
Hello guys!

I love Inertia and I love making applications with Inertia.

Right now I'm working on a project with many Policies and Gates and I'm thinking if there is a better way to do it.

Currently for class based policy methods I'm always appending **can** attribute to the User. For model based policy methods I'm appending **can** attribute for every Model.

This is the Inertia's way. 

### **The problem**

Currently I have model **Ticket** with 20+ policy methods and some of them are required to load additional relationships in order to check the policy result. 

This results to many loaded models and honestly I don't like it. 

Is there any way of making Inertia smarter and lazy load only the policy methods needed for the next page ? 

What do you think?

**Comment by ajnsn** [comment]
Hey @Bojmaliev 
I often see implementations with Inertia's shared data to maintain the auth user with the overall can-attributes while returning model based can-attributes together with the record response. So in general that's what you are already doing.

The performance/relationships issue you describe is something that should also happen in any application, e.g. when using Laravel with blade compoments you should encounter the same issue. So while I unfortunately could not provide any instant solution for you, I recommend reviewing your policy concept and try to optimize the queries. If you are using Laravel, you could also activate `Model::preventLazyLoading` to [identify N+1 query problems](https://laravel-news.com/disable-eloquent-lazy-loading-during-development).

I hope this helps a bit to get in the right direction!

**Reply by bojmaliev** [reply]
Hey @ajnsn,

Thank you for your opinion.

You are right. That's what I'm doing. The only key difference between Inertia and Blade is that when I'm working with Blade, the server is evaluating only the Policy Methods that are needed for the current request.

When I'm working with Inertia or any other SPA, I need to tell the server to evaluate all the user-can attributes, no matter if they will be used in the next page or not.

I'm ok when the server need to load relationships to check if a policy should return true or false but I don't like when it does when I don't need that result in that request.

The only idea that came to me is to generate this user-can attributes separately for every request But I think my code will get messy.

Or, if it was possible to define each of these attributes as Lazy loaded page props and then for each page somehow Inertia to know which one to call for partial.

**Reply by ajnsn** [reply]
The concept behind would be to return the can-attribute with every response but dedicated to this record/reponse.
Just curious but If you return a record based can-attribute, like [shown here](https://inertiajs.com/authorization) it should keep your policy evaluation low (comparable to the blade example)? This should also work in lists/multiple records pretty well unless you have a deep nested policy structure? 

Can you show some code how you pass your policy evaluations to the frontend?

**Reply by bojmaliev** [reply]
This is model-based:

```
class Ticket extends Model
{
 protected $appends = ['can'];
 public function getCanAttribute(): array
    {
        return [
            Permission::assign => Gate::check(Permission::assign, $this),
            Permission::view => Gate::check(Permission::view, $this),
            Permission::refuse => Gate::check(Permission::refuse, $this),
            Permission::start => Gate::check(Permission::start, $this),
            Permission::finish => Gate::check(Permission::finish, $this),
            Permission::confirm => Gate::check(Permission::confirm, $this),
            Permission::return => Gate::check(Permission::return, $this),
            Permission::cancel => Gate::check(Permission::cancel, $this),
            Permission::createChildren => Gate::check(Permission::createChildren, $this),
            Permission::comment => Gate::check(Permission::comment, $this),
            Permission::flag => Gate::check(Permission::flag, $this),
            Permission::priority => Gate::check(Permission::priority, $this),
        ];
    }
}

```
And I Have user-based:


```
class User extends Authenticatable
{

 public function getCanAttribute(): array
    {
        return [
            Permission::configs => [
                Permission::viewAny => $this->can(Permission::viewAny, Config::class),
                Permission::edit => $this->can(Permission::edit, Config::class),
            ],
            Permission::places => [
                Permission::viewAny => $this->can(Permission::viewAny, Place::class),
                Permission::create => $this->can(Permission::create, Place::class),
                Permission::edit => $this->can(Permission::edit, Place::class),
            ],
      ]
}
}
```


And then in HandleInertiaRequests


```
 public function share(Request $request)
    {
        $user = $request->user();
        return array_merge(parent::share($request), [
            'authUser' => fn() => $user?->append('can'),
        ]);
    }
```

This works fine but I know this is not ideal. 

I'm afraid if I use [this ](https://inertiajs.com/authorization) all my methods will become messy. Also how can I append to paginated results?

**Reply by ajnsn** [reply]
Interesting! But probably very expensive, as you already noticed. You could take a look at [Parital reloads](https://inertiajs.com/partial-reloads#lazy-data-evaluation) but in the end I would not recommend following this shared data approach. 

Instead you could take advantage of Laravel's native [Authorization policies](https://laravel.com/docs/8.x/authorization#creating-policies) and then be able to just e.g. `auth()->user()->can('update', $model)`. 

Furthermore take a look at Laravel's [API resources](https://laravel.com/docs/8.x/eloquent-resources). They can ease up the transformation layer and can also be used inside `Inertia::render`. Try to achieve the same as the [Authorization example](https://inertiajs.com/authorization) does. It works pretty well. 

Hope this helps :)

**Reply by bojmaliev** [reply]
Thank you for your advice and thanks for having this discussion with me.

The thing I don't like for API resources is that I need to create separate one for each response, but I will give it a try.

Again, If just there was a way to tell Inertia: Hey please lazy load me these 5 policy methods for my next page, I will need them.

**Reply by ajnsn** [reply]
You're welcome! API Resources are optional. Just follow the Policy path, it's this simple Laravel way of model policies. If you want to keep your existing approach, you could also think about [caching them](https://laravel.com/docs/8.x/cache#retrieve-store).

**Reply by bojmaliev** [reply]
I'm already doing everything Laravel recommends, Just I'm not sure what's the best way of providing them to the front-end. 
Caching is also a good idea. Thanks
